### PR TITLE
Dev/feng/goalkeeper controller

### PR DESCRIPTION
**Pull request by fengfengFinn**
Hey, guys. Here is a draft verion about a method of persuiting a ball. There still somewhere need to modify. I will try to complete this and give a demo verion asap.

### CODE REVIEW

File: GameEngine/robot_model.py [hunk 1]
```diff
@@ -91,6 +92,14 @@ def get_position_components_wcs(self) -> (float, float):
         """
         return self._convert_EFCS_to_field_CS(self._x_pos_EFCS, self._y_pos_EFCS)
 
+    def get_point_angle_wcs(self):
+
+        return self.pointing_angle
```
Comment (szgula) [inline GameEngine/robot_model.py:97]: the self.pointing_angle is in EFCS (find description in the RobotModel. __init__ method)

so if you require the pointing in EFCS you can access this parameter directly: 
`my_robot = RobotBasicModel()
angle = my_robot.pointing_angle`
 (WCS stands for World Coordinate System, the CS shared by all non players related components)

Anyway - keep this function for WCS - but take into consideration the `self._cord_system_rot` parameter

https://github.com/szgula/UofG_Robotics_TDP/blob/0dd4c3637a14d72422c78d8c6a44a278d3da4969/GameEngine/robot_model.py#L24

File: GameEngine/robot_model.py [hunk 2]
```diff
@@ -91,6 +92,14 @@ def get_position_components_wcs(self) -> (float, float):
...
+
+    def get_axis_len(self):
+
+        return self._axis_len
```
Comment (szgula) [inline GameEngine/robot_model.py:101]: all parameters with underscore before name (`_my_param`) should be internal, if you require it outside the object - rename it to `self.axis_len` (remove first underscore), get value directly (`my_object.axis_len`) and remove getter - it is not Java :p

File: GameEngine/visualizer.py [hunk 3]
```diff
@@ -23,6 +23,8 @@ def __init__(self, simulator: None, number_of_players: int = 2, field_size: tupl
         self._temp_initialized_data = False
         self._data_provided = False
         self._data_provided_buffer = ()
+        self.fps = 500
```
Comment (szgula) [inline GameEngine/visualizer.py:26]: fps = 500? This is maximum fps we allow the game - right? is it just happy number or you selected it for some reason? (probably we would be fine with much smaller value)

also, we probably should synchronise it with the `dt` (delta time) in the game simulator

File: Robots/goalkeeper_controller.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    def __init__(self):
+        self._goalkeeper = None
+        self._ball = None
+        self._lookhead = 10
+
+    def get_action(self, goal) -> tuple:
+        if goal == Goal.ChaseBall:
+            lookahead = self._ball.get_position()
+
```
Comment (szgula) [inline Robots/goalkeeper_controller.py:18]: line breaks should be between different blocks of the code, between methods in class, double lines between imports and other code and after the class 

Take a look on PEP8 -> for example here: [PEP8 RealPython](https://realpython.com/python-pep8/)

File: Robots/goalkeeper_controller.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import numpy as np
+import math
+from GameEngine.ball_model import BallModel
+from GameEngine.robot_model import RobotModel
+from Robots.robot_control import Robot, Goal
+
+
+class GoalkeeperController(Robot):
+
+    def __init__(self):
```
Comment (szgula) [inline Robots/goalkeeper_controller.py:10]: reference to the robot model could be passed in object initialisation, probably same with ball (not in `accrue_sensors_data`)

..Thus, .I'm not sure if this is right approach; so you can keep current structure and we can get back to it another time

File: GameEngine/game_master.py [hunk 6]
```diff
@@ -1,10 +1,17 @@
+import sys
```
Comment (szgula) [inline GameEngine/game_master.py:1]: sys and pygame is only needed in `__name__ == "__main__"` how about we move import inside the `if`?

**Review by szgula — CHANGES_REQUESTED** [review]
- [x] Take a look at the comments 
- [ ] Do rebase to resolve merge conflicts

**Comment by szgula** [comment]
@fengfengFinn Nice work :) I added some comments, you do not need to agree ;) some of them are functional - other just to give you some feedback about python coding style and conventions

**Comment by szgula** [comment]
Also, you need to resolve conflicts before you merge your changes - probably easiest way it to do **_rebase_** of your branch - if you get any problems with that - let me know, I will try to support you :)

File: GameEngine/robot_model.py (hunk 2 above)
Comment (fengfengFinn) [reply GameEngine/robot_model.py]: Gotcha

**Review by szgula — APPROVED** [review]
This is good :)

- [ ] was it expected that you don't use this controller in game_master now?
- [ ] ask @OmarJabri7 or @Yifan-Xie to run this on another computer

When both above are positive - go for merge! GOOD WORK! 🥇

**Comment by szgula** [comment]
:+1:

**Pull request merged**